src/iam-token.js

```diff
@@ -0,0 +1,147 @@
+    const apiKeyRef = objectPath.get(iam, 'apiKeyRef');
+
+    if (typeof apiKeyAlpha1 == 'string') {
+      apiKey = apiKeyAlpha1;
+    } else if (typeof apiKeyStr == 'string') {
+      apiKey = apiKeyStr;
+    } else if (typeof apiKeyAlpha1 == 'object') {
+      let secretName = objectPath.get(apiKeyAlpha1, 'valueFrom.secretKeyRef.name');
+      let secretNamespace = objectPath.get(apiKeyAlpha1, 'valueFrom.secretKeyRef.namespace', namespace);
+      let secretKey = objectPath.get(apiKeyAlpha1, 'valueFrom.secretKeyRef.key');
+      apiKey = await this._getSecretData(secretName, secretKey, secretNamespace, kubeResourceMeta, namespace);
+    } else if (typeof apiKeyRef == 'object') {
+      let secretName = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.name');
+      let secretNamespace = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.namespace', namespace);
+      let secretKey = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.key');
+      apiKey = await this._getSecretData(secretName, secretKey, secretNamespace, kubeResourceMeta, namespace);
+    }
+    if (!apiKey) {
+      return Promise.reject('Failed to find valid apikey to authenticate against iam');
+    }
+
+    let res = await this._requestToken(iam, apiKey);
+    return res.access_token;
+  }
+
+
+  async _requestToken(iam, apiKey) {
+    let token;
+
+    const apiKeyHash = hash(apiKey, { algorithm: 'shake256' });
+    // Check if the cache has the token. If it does have it, copy it into token
+    // for further processing
+    if (objectPath.has(this.s3TokenCache, [apiKeyHash])) {
+      // fetch cached token
+      token = objectPath.get(this.s3TokenCache, [apiKeyHash]);
+    }
+
+    // if we are able to fetch the token from the cache and it is not a Promise,
+    // then we need to check the expiry. If it is not expiring in the next 120 s
+    // return the token
+    if (!(token instanceof Promise)) {
+      if (token !== undefined) {
+        const expires = objectPath.get(token, 'expiration', 0); // expiration: time since epoch in seconds
+        // re-use cached token as long as we are more than 2 minutes away from expiring.
+        if (Date.now() < (expires - 120) * 1000) {
+          return token;
+        }
+      }
+
+      // if we get to this point the token is either undefined or it is stale data.
+      // either ways we need to connect to iam auth and get a new token
+      // create an async function that will wait for _getToken to return a token.
+      // since we are not using await, the control will return immediately and the
+      // return value of the tokenFunction is going to be a promise.
+      // we are going to cache this promise into the s3TokenCache and
+      // wait for it.
+      let tokenFunction = async (iam, apiKeyHash, apiKey) => {
+        let ret = await this._getToken(iam, apiKeyHash, apiKey);
+        return ret;
+      };
+      token = tokenFunction(iam, apiKeyHash, apiKey);
```

Comment:
If later on you'll get the token by assuming that `_getToken` sets it into the global cache and just retrieving it from the cache, as at line 102:
```
        await token;
        return objectPath.get(this.s3TokenCache, [apiKeyHash]);
```
then there's no reason for tokenFunction to return the value.  And I think no reason to define an inline token function at all.  Can't you just do this?
```suggestion
      token = this._getToken(iam, apiKeyHash, apiKey);
```
Alternately, you could remove the side-effect of setting the cache value from `_getToken` and set it explicitly either in the `tokenFunction`, or returning a value from `tokenFunction` and then setting the cache value.

Reply:
done